### PR TITLE
Fix regridding issue where target grid is more than 2D

### LINKED ISSUE #2085: Rechunking to keep chunk size normal may fail
For example, when running recipe_autoassess_stratosphere.yml from ESMValTool with the current `main` branch, the run fails to regrid dataset MPI-ESM-LR with:
```
ValueError: Chunks do not add up to shape. Got chunks=((1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1), (1, 1, 1, 1, 1, 1, 1), (12,), (19,)), shape=(132, 7, 96, 192)
```

Full log: [main_log.txt](https://github.com/ESMValGroup/ESMValCore/files/11684934/main_log.txt)

### PR DESCRIPTION
**Pull request by bouweandela**
## Description

In #968 a rechunking step was introduced if the target grid resolution was considerably higher than the source grid. However, it failed to take into account that the target grid could have more dimensions than just latitude and longitude. This pull request fixes the issue.

Closes #2085

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

File: esmvalcore/preprocessor/_regrid.py [hunk 1]
```diff
@@ -681,22 +681,36 @@ def regrid(cube, target_grid, scheme, lat_offset=True, lon_offset=True):
...
+    if lon_coord.ndim != 1 or lat_coord.ndim != 1:
+        # This function only supports 1D lat/lon coordinates.
+        return cube
+
+    lon_dim, = target_grid.coord_dims(lon_coord)
+    lat_dim, = target_grid.coord_dims(lat_coord)
+    grid_indices = sorted((lon_dim, lat_dim))
+    target_grid_shape = tuple(target_grid.shape[i] for i in grid_indices)
+
+    if 2 * np.prod(cube.shape[-2:]) > np.prod(target_grid_shape):
```
Comment (zklaus) [inline esmvalcore/preprocessor/_regrid.py:704]: All the new code above seems to be here to allow for the target_grid to have x and y dimensions at arbitrary places. Is my reading correct? In that case, should we afford cube the same possibility, or perhaps it's overkill also for target_grid?

**Review by zklaus — COMMENTED** [review]
Just one little question. Otherwise looks good.

File: esmvalcore/preprocessor/_regrid.py (hunk 1 above)
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_regrid.py:704]: that's a good point! alas, this was in the ancient regrid module eons ago, so why not?

Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid.py:704]: Currently, the target grid can be an arbitrary NetCDF file, while `cube` is CMORized, so I would expect it to have lat/lon as the last two dimensions.

File: tests/unit/preprocessor/_regrid/test_regrid.py [hunk 2]
```diff
@@ -380,5 +407,36 @@ def test_no_rechunk_non_lazy():
...
+    )
+    tgt_grid.add_aux_coord(
+        iris.coords.AuxCoord(
+            np.broadcast_to(lon_points.reshape(1, -1), tgt_grid_dims),
+            standard_name='longitude',
+        ),
+        (0, 1),
+    )
+    result = _rechunk(cube, tgt_grid)
+    assert result.core_data().chunks == cube.core_data().chunks
```
[the remark this answers is not in the document]
Comment (valeriupredoi) [reply tests/unit/preprocessor/_regrid/test_regrid.py]: yay!

File: esmvalcore/preprocessor/_regrid.py [hunk 3]
```diff
@@ -681,22 +681,36 @@ def regrid(cube, target_grid, scheme, lat_offset=True, lon_offset=True):
...
     """Re-chunk cube with optimal chunk sizes for target grid."""
     if not cube.has_lazy_data() or cube.ndim < 3:
         # Only rechunk lazy multidimensional data
         return cube
 
-    if 2 * np.prod(cube.shape[-2:]) > np.prod(target_grid.shape):
+    lon_coord = target_grid.coord(axis='X')
+    lat_coord = target_grid.coord(axis='Y')
+    if lon_coord.ndim != 1 or lat_coord.ndim != 1:
+        # This function only supports 1D lat/lon coordinates.
```
[the remark this answers is not in the document]
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_regrid.py:696]: can't regrid cubes with 2D coordinates so go home, or something?

**Review by valeriupredoi — APPROVED** [review]
cheers, Bouwe :beer:

File: esmvalcore/preprocessor/_regrid.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid.py:696]: We can regrid cubes with 2D coordinates, but those are handled by the code above. Except for the case where the source cube has 1D coordinates and the target cube has 2D coordinates. That case is not handled at all, but I want to defer throwing an exception for that to the iris regridder that is called after rechunk.

[the remark this answers is not in the document]
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_regrid.py:696]: aha gotcha, cheers, bud

**Pull request merged**